### PR TITLE
Use ActionsDAG in KeyCondition

### PR DESCRIPTION
**Pull request by KochetovNicolai**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Change the way how PK is analyzed for MergeTree.

### CODE REVIEW

**Review by novikd — APPROVED** [review]
In general LGTM, but maybe you want to randomly change `query_plan_optimize_primary_key` value in stress tests.

**Comment by KochetovNicolai** [comment]
@novikd I want to have `query_plan_optimize_primary_key` always enabled, and then remove `enable_optimize_predicate_expression` eventually :)

File: src/QueryPipeline/QueryPipelineBuilder.h [hunk 1]
```diff
@@ -94,6 +94,9 @@ class QueryPipelineBuilder
     /// Changes the number of output ports if needed. Adds ResizeTransform.
     void resize(size_t num_streams, bool force = false, bool strict = false);
 
+    /// Concat some ports to have no more then size outputs.
+    void narrow(size_t size);
```
[the remark this answers is not in the document]
Comment (KochetovNicolai) [reply src/QueryPipeline/QueryPipelineBuilder.h:98]: Yes, I will add comment

**Comment by den-crane** [comment]
it introduced a bug https://github.com/ClickHouse/ClickHouse/issues/40599

**Pull request merged**